cpp/src/arrow/util/small_vector.h

```diff
@@ -0,0 +1,115 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <initializer_list>
+#include <utility>
+#include <vector>
+
+#include "arrow/util/variant.h"
+
+namespace arrow {
+namespace util {
+
+/// A minimal vector-like container which holds up to a single element without allocation.
```

Comment:
That's neat! Can you add tests for this?
(also, it would be nice if you could also parameter the class with the number of static elements)

Reply:
I can add tests.
If we want a full `small_vector` then we should probably vendor one. In any case I'll leave that for a follow up JIRA

Reply:
On consideration, I'll remove `small_vector` for now. Follow up to re-add https://issues.apache.org/jira/browse/ARROW-8022